libraries/lib-music-information-retrieval/StftFrameProvider.cpp

```diff
@@ -0,0 +1,91 @@
+#include "StftFrameProvider.h"
+#include "MirAudioReader.h"
+#include "MirUtils.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <numeric>
+
+namespace MIR
+{
+namespace
+{
+constexpr auto twoPi = 2 * 3.14159265358979323846;
+
+int GetFrameSize(int sampleRate)
+{
+   // 2048 frame size for sample rate 44.1kHz
+   return 1 << (11 + (int)std::round(std::log2(sampleRate / 44100.)));
+}
+
+double GetHopSize(int sampleRate, long long numSamples)
+{
+   // Aim for a hop size closest to 10ms, yet dividing `numSamples` to a power
+   // of two. This will spare us the need for resampling when we need to get the
+   // autocorrelation of the ODF using an FFT.
+   const auto idealHopSize = 0.01 * sampleRate;
+   const auto numFrames =
+      1 << (int)std::round(std::log2(numSamples / idealHopSize));
+   return 1. * numSamples / numFrames;
+}
+} // namespace
+
+StftFrameProvider::StftFrameProvider(const MirAudioReader& audio)
+    : mAudio { audio }
+    , mFftSize { GetFrameSize(audio.GetSampleRate()) }
+    , mHopSize { GetHopSize(audio.GetSampleRate(), audio.GetNumSamples()) }
+    , mWindow { GetNormalizedHann(mFftSize) }
+    , mNumFrames { static_cast<int>(
+         std::round(audio.GetNumSamples() / mHopSize)) }
+    , mNumSamples { audio.GetNumSamples() }
+{
+   assert(IsPowOfTwo(mNumFrames));
+}
+
+bool StftFrameProvider::GetNextFrame(std::vector<float>& frame)
+{
+   if (mNumFramesProvided >= mNumFrames)
+      return false;
+   frame.resize(mFftSize);
+   const int firstReadPosition = mHopSize - mFftSize;
+   int start = std::round(firstReadPosition + mNumFramesProvided * mHopSize);
+   while (start < 0)
+      start += mNumSamples;
+   const auto end = std::min<long long>(start + mFftSize, mNumSamples);
+   const auto numToRead = end - start;
+   mAudio.ReadFloats(frame.data(), start, numToRead);
+   // It's not impossible that some user drops a file so short that `mFftSize >
+   // mNumSamples`. In that case we won't be returning a meaningful
+   // STFT, but that's a use case we're not interested in. We just need to make
+   // sure we don't crash.
```

Comment:
To treat the short file boundary case really fully, should there be a while loop below?  With a careful test that the loop makes progress each time?  Or should you be sure to zero-fill any uninitialized part remaining after the second read?  (Which would be a consequence of first resizing the buffer to zero above)

Reply:
Yes, the perfect circular read-out would need a while loop. It probably makes implementation easier not to care for this use case, which would be too short to be a musical loop anyway.
I think initialization to 0 would only improve situations where the underlying file is less than twice the FFT size, but it's cheap and easy, so will do.